web/src/compose_reply.js

```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+        left: selection_bounds.left - message_content_rect.left,
+        width: selection_bounds.right - selection_bounds.left,
+        height: selection_bounds.bottom - selection_bounds.top,
+    };
+    const reference_div = document.createElement("div");
+    reference_div.style.position = "absolute";
+    reference_div.style.top = `${relative_rect.top}px`;
+    reference_div.style.left = `${relative_rect.left}px`;
+    reference_div.style.width = `${relative_rect.width}px`;
+    reference_div.style.height = `${relative_rect.height}px`;
+    // Allow clicks to go through.
+    reference_div.style.pointerEvents = "none";
+    message_content_div.append(reference_div);
+    return reference_div;
+}
+
+function show_quote_and_reply_popover(reference_element, message_id, selected_message_content_raw) {
+    popover_menus.toggle_popover_menu(reference_element, {
+        popperOptions: {
+            modifiers: [
+                {
+                    name: "flip",
+                    options: {
+                        fallbackPlacements: ["bottom", "left", "right"],
+                    },
+                },
+            ],
+        },
+        // Add a delay to separate `hideOnClick` and `onShow` so that
+        // `onShow` is called after `hideOnClick`.
+        // See https://github.com/atomiks/tippyjs/issues/230 for more details.
+        delay: [100, 0],
+        onShow(instance) {
+            popover_menus.popover_instances.quote = instance;
+            instance.setContent(parse_html(render_quote_and_reply_popover()));
+            return true;
+        },
+        onHidden(instance) {
+            instance.destroy();
+            reference_element.remove();
+            popover_menus.popover_instances.quote = undefined;
+        },
+        onMount(instance) {
+            const $popper = $(instance.popper);
+            $popper.on("click", ".quote_and_reply", (e) => {
+                quote_and_reply({
+                    message_id,
+                    selected_message_content_raw,
+                });
+                instance.hide();
+                e.preventDefault();
+                e.stopPropagation();
+            });
+        },
+    });
+}
+
+function suggest_quoting_message_selection(selection = window.getSelection()) {
+    if (!selection?.toString()) {
+        return;
+    }
+    const {start_id, end_id} = copy_and_paste.analyze_selection(selection);
+    // If selection spans multiple messages, do nothing.
+    if (start_id !== end_id) {
+        return;
+    }
+    const {selected_message_content_raw, selection_bounds} = get_message_selection(selection);
+
+    if (selected_message_content_raw) {
+        const message_id = start_id;
+        const reference_element = create_popover_reference_in_message(
+            document.querySelector(
+                `.focused-message-list [data-message-id="${message_id}"] .message_content`,
+            ),
+            selection_bounds,
+        );
```

Comment:
Would it be cleaner to move the `create_popover_reference_in_message` call into `show_quote_and_reply_popover`, and just pass `message_id, selection_bounds, selected_message_content_raw` instead?